[exporter/splunkhec] Flaky Test - TestReceiveTracesBatches

Observed here: https://github.com/open-telemetry/opentelemetry-collector-contrib/runs/6969904966?check_suite_focus=true#step:7:4748

```
=== RUN   TestReceiveTracesBatches/1_trace_event_per_payload_(configured_max_content_length_is_same_as_event_size)
    client_test.go:441: 
        	Error Trace:	client_test.go:441
        	Error:      	"{\"time\":2,\"host\":\"unknown\",\"event\":{\"trace_id\":\"01010101010101010101010101010101\",\"span_id\":\"0000000000000001\",\"parent_span_id\":\"\",\"name\":\"root\",\"end_time\":3000000000,\"kind\":\"SPAN_KIND_UNSPECIFIED\",\"status\":{\"message\":\"\",\"code\":\"STATUS_CODE_UNSET\"},\"start_time\":2000000000},\"fields\":{\"resource\":\"R1\"}}" does not contain "\"start_time\":1"
        	Test:       	TestReceiveTracesBatches/1_trace_event_per_payload_(configured_max_content_length_is_same_as_event_size)
    client_test.go:441: 
        	Error Trace:	client_test.go:441
        	Error:      	"{\"time\":3,\"host\":\"unknown\",\"event\":{\"trace_id\":\"01010101010101010101010101010101\",\"span_id\":\"0000000000000001\",\"parent_span_id\":\"0102030405060708\",\"name\":\"root\",\"end_time\":4000000000,\"kind\":\"SPAN_KIND_UNSPECIFIED\",\"status\":{\"message\":\"ok\",\"code\":\"STATUS_CODE_OK\"},\"start_time\":3000000000},\"fields\":{\"resource\":\"R1\"}}" does not contain "\"start_time\":2"
        	Test:       	TestReceiveTracesBatches/1_trace_event_per_payload_(configured_max_content_length_is_same_as_event_size)
    client_test.go:441: 
        	Error Trace:	client_test.go:441
        	Error:      	"{\"time\":4,\"host\":\"unknown\",\"event\":{\"trace_id\":\"01010101010101010101010101010101\",\"span_id\":\"0000000000000001\",\"parent_span_id\":\"\",\"name\":\"root\",\"end_time\":5000000000,\"kind\":\"SPAN_KIND_UNSPECIFIED\",\"status\":{\"message\":\"\",\"code\":\"STATUS_CODE_UNSET\"},\"start_time\":4000000000},\"fields\":{\"resource\":\"R1\"}}" does not contain "\"start_time\":3"
        	Test:       	TestReceiveTracesBatches/1_trace_event_per_payload_(configured_max_content_length_is_same_as_event_size)
    client_test.go:441: 
        	Error Trace:	client_test.go:441
        	Error:      	"{\"time\":1,\"host\":\"unknown\",\"event\":{\"trace_id\":\"01010101010101010101010101010101\",\"span_id\":\"0000000000000001\",\"parent_span_id\":\"0102030405060708\",\"name\":\"root\",\"end_time\":2000000000,\"kind\":\"SPAN_KIND_UNSPECIFIED\",\"status\":{\"message\":\"ok\",\"code\":\"STATUS_CODE_OK\"},\"start_time\":1000000000},\"fields\":{\"resource\":\"R1\"}}" does not contain "\"start_time\":4"
        	Test:       	TestReceiveTracesBatches/1_trace_event_per_payload_(configured_max_content_length_is_same_as_event_size)
=== RUN   TestReceiveTracesBatches/2_trace_events_per_payload_(configured_max_content_length_is_twice_event_size)
=== RUN   TestReceiveTracesBatches/1_compressed_batch_of_2037_bytes,_make_sure_the_event_size_is_more_than_minCompressionLen=1500_to_trigger_compression
=== RUN   TestReceiveTracesBatches/2_compressed_batches_-_1832_bytes_each,_make_sure_the_log_size_is_more_than_minCompressionLen=1500_to_trigger_compression
--- FAIL: TestReceiveTracesBatches (5.08s)
    --- PASS: TestReceiveTracesBatches/all_trace_events_in_payload_when_max_content_length_unknown_(configured_max_content_length_0) (1.02s)
    --- FAIL: TestReceiveTracesBatches/1_trace_event_per_payload_(configured_max_content_length_is_same_as_event_size) (1.01s)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -55,6 +55,7 @@
 - `googlecloudexporter`: Fix (self-obs) point_count metric calculation, concurrent map write panic, and dropped log attributes (#11051)
 - `signalfxexporter`: Event Type is a required field, if not set, set it to `unknown` to prevent signalfx ingest from dropping it (#11121)
 - `prometheusreceiver`: validate that combined metric points (e.g. histograms) have the same timestamp (#9385)
+- `splunkhecexporter`: Fix flaky test when exporting traces (#11418)
 
 ## v0.53.0
 

--- a/exporter/splunkhecexporter/client_test.go
+++ b/exporter/splunkhecexporter/client_test.go
@@ -26,6 +26,8 @@ import (
 	"net"
 	"net/http"
 	"net/url"
+	"regexp"
+	"sort"
 	"sync"
 	"testing"
 	"time"
@@ -44,6 +46,8 @@ import (
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/splunk"
 )
+
+var requestTimeRegex = regexp.MustCompile(`time":(\d+)`)
 
 type testRoundTripper func(req *http.Request) *http.Response
 
@@ -277,6 +281,18 @@ func runTraceExport(testConfig *Config, traces ptrace.Traces, t *testing.T) ([]r
 			if len(requests) == 0 {
 				err = errors.New("timeout")
 			}
+
+			// sort the requests according to the traces we received, reordering them so we can assert on their size.
+			sort.Slice(requests, func(i, j int) bool {
+				imatch := requestTimeRegex.FindSubmatch(requests[i].body)
+				jmatch := requestTimeRegex.FindSubmatch(requests[j].body)
+				// no matches mean it's compressed, just leave as is
+				if len(imatch) == 0 {
+					return i < j
+				}
+				return string(imatch[1]) <= string(jmatch[1])
+			})
+
 			return requests, err
 		}
 	}

